<a href="https://colab.research.google.com/github/GUNAPILLCO/neural_profit/blob/main/3_dise%C3%B1o_entrenamiento_evaluacion/3_2_Generacion_X_y.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 3.2. Generación de X e y para modelos

## 0. Clonado de Repositorio, instalación de librería e importación.

In [1]:
#Clonamos el repo
#LINK DE REPOSITORIO: https://github.com/GUNAPILLCO/neural_profit

!git clone https://github.com/GUNAPILLCO/neural_profit.git

fatal: destination path 'neural_profit' already exists and is not an empty directory.


In [2]:
# !{sys.executable} -m pip install -q pandas_market_calendars  # Solo si usás horarios de mercados
!{sys.executable} -m pip install -q ta
print("Librerías instaladas: ta")

/bin/bash: line 1: {sys.executable}: command not found
Librerías instaladas: ta


In [3]:
import sys
import warnings
warnings.filterwarnings('ignore')

# Utilidades del sistema y fechas
import os
import glob
import requests
from datetime import datetime, timedelta

# Procesamiento de datos
import pandas as pd
import numpy as np

# Visualización
import matplotlib.pyplot as plt
import seaborn as sns
from tabulate import tabulate

# Análisis técnico
import ta
from ta.momentum import StochasticOscillator, ROCIndicator
from ta.volatility import BollingerBands, AverageTrueRange

# Estadística
from scipy.stats import spearmanr

#que es?
#from tqdm.notebook import tqdm
from tqdm import tqdm
# Modelos ML
#from xgboost import XGBRegressor
#from sklearn.metrics import mean_squared_error, r2_score

# Calendario de mercados (descomentar si lo necesitás)
# import pandas_market_calendars as mcal

## 1. Carga de datasets train, valid y test.

In [4]:
def load_df():
    """
    Función para cargar un archivo Parquet desde el repositorio clonado
    """
    # Definir la URL del archivo Parquet en GitHub
    df_path_mnq = '/content/neural_profit/3_diseño_entrenamiento_evaluacion/mnq_model.parquet'
    df_path_factores = '/content/neural_profit/3_diseño_entrenamiento_evaluacion/df_factores.parquet'
    df_path_train = '/content/neural_profit/3_diseño_entrenamiento_evaluacion/mnq_train.parquet'
    df_path_valid = '/content/neural_profit/3_diseño_entrenamiento_evaluacion/mnq_valid.parquet'
    df_path_test = '/content/neural_profit/3_diseño_entrenamiento_evaluacion/mnq_test.parquet'

    # Leer el archivo Parquet y cargarlo en un DataFrame
    df_model = pd.read_parquet(df_path_mnq)
    df_factores = pd.read_parquet (df_path_factores)
    df_train = pd.read_parquet(df_path_train)
    df_valid = pd.read_parquet(df_path_valid)
    df_test = pd.read_parquet(df_path_test)

    return df_model, df_factores, df_train, df_valid, df_test

In [5]:
mnq_model, indicadores_tecnicos, mnq_train, mnq_valid, mnq_test = load_df()

## 1.1. Información de los datasets

In [6]:
def info_dataset (df): # Contar valores únicos en la columna 'date'
  num_dias = df['date'].nunique()
  print(f"Cantidad de días: {num_dias}")

  # Filtrar valores válidos
  validos_por_dia = df.dropna(subset=['target_return_30']).groupby('date').size()

  # Calcular el promedio
  promedio_por_fecha = validos_por_dia.mean()
  print(f"Valores por día: {int(promedio_por_fecha)}")

  primer_hora = df.index[0].strftime('%H:%M')
  ultima_hora = df.index[-1].strftime('%H:%M')
  zona_horaria = df.index[0].tzinfo
  print(f"Hora diaria de inicio {primer_hora}")
  print(f"Hora diaria de final {ultima_hora}")
  print(f"Zona horaria: {zona_horaria}")

In [7]:
info_dataset(mnq_train)

Cantidad de días: 917
Valores por día: 361
Hora diaria de inicio 09:30
Hora diaria de final 15:30
Zona horaria: America/New_York


In [8]:
info_dataset(mnq_valid)

Cantidad de días: 197
Valores por día: 361
Hora diaria de inicio 09:30
Hora diaria de final 15:30
Zona horaria: America/New_York


In [9]:
info_dataset(mnq_test)

Cantidad de días: 197
Valores por día: 361
Hora diaria de inicio 09:30
Hora diaria de final 15:30
Zona horaria: America/New_York


## 2. Generación de X e y

In [15]:
target_column = "target_return_30"
features = mnq_model.columns.tolist()
features.remove( target_column)
features.remove( 'date')
window_size = 60

In [19]:
len(features)

21

In [20]:
#Función para generar ventanas y vectorizarlas
def generar_ventanas(df, features, target_col, window_size):
    X, y = [], []
    for fecha, grupo in tqdm(df.groupby("date"), desc="Procesando días"):
        grupo = grupo.reset_index(drop=True)
        for i in range(len(grupo) - window_size):
            ventana = grupo.loc[i:i+window_size-1, features]
            if ventana.isnull().any().any():
                continue
            vector = ventana.values.flatten()
            target = grupo.loc[i+window_size-1, target_col]
            X.append(vector)
            y.append(target)
    return np.array(X), np.array(y)

In [21]:
#Generar X y y para train/valid/test
print('Generando X_train e y_train: \n')
X_train, y_train = generar_ventanas(mnq_train, features, target_column, window_size)

#print('Generando X_valid e y_valid: \n')
#X_valid, y_valid = generar_ventanas(mnq_valid, features, target_column, window_size)3

#print('Generando X_test e y_test: \n')
#X_test, y_test = generar_ventanas(mnq_test, features, target_column, window_size)


Generando X_train e y_train: 



Procesando días: 100%|██████████| 917/917 [04:47<00:00,  3.19it/s]


In [22]:
X_train.shape

(276017, 1260)

In [23]:
y_train.shape

(276017,)

X_train: 276017 ventanitas, cada una aplanada en 1,260 números (21 features × 60 min).

y_train: 276,017 valores que son los retornos a 30 min después de cada ventanita.

In [26]:
print('Generando X_valid e y_valid: \n')
X_valid, y_valid = generar_ventanas(mnq_valid, features, target_column, window_size)

print('Generando X_test e y_test: \n')
X_test, y_test = generar_ventanas(mnq_test, features, target_column, window_size)

Generando X_valid e y_valid: 



Procesando días: 100%|██████████| 197/197 [00:58<00:00,  3.39it/s]


Generando X_test e y_test: 



Procesando días: 100%|██████████| 197/197 [00:58<00:00,  3.35it/s]


##Guardar

In [28]:
# Definir rutas dentro de tu repo
ruta_X_train = '/content/neural_profit/3_diseño_entrenamiento_evaluacion/X_train.npy'
ruta_y_train = '/content/neural_profit/3_diseño_entrenamiento_evaluacion/y_train.npy'

ruta_X_valid = '/content/neural_profit/3_diseño_entrenamiento_evaluacion/X_valid.npy'
ruta_y_valid = '/content/neural_profit/3_diseño_entrenamiento_evaluacion/y_valid.npy'

ruta_X_test = '/content/neural_profit/3_diseño_entrenamiento_evaluacion/X_test.npy'
ruta_y_test = '/content/neural_profit/3_diseño_entrenamiento_evaluacion/y_test.npy'

# Guardar arrays en disco
np.save(ruta_X_train, X_train)
np.save(ruta_y_train, y_train)

np.save(ruta_X_valid, X_valid)
np.save(ruta_y_valid, y_valid)

np.save(ruta_X_test, X_test)
np.save(ruta_y_test, y_test)

print("✅ Arrays X_*/y_* guardados en el repositorio como .npy")

✅ Arrays X_*/y_* guardados en el repositorio como .npy


In [29]:
# Cambiar al directorio del repo
%cd /content/neural_profit

token = ""  # tu token guardado localmente

!git config --global user.email "gusunapillco@gmail.com"
!git config --global user.name "GUNAPILLCO"
!git add .
!git commit -m "Ventanas X_train/X_valid/X_test y y_train/y_valid/y_test guardadas como .npy"
!git push https://GUNAPILLCO:{token}@github.com/GUNAPILLCO/neural_profit.git

print("✅ Arrays subidos al repositorio")

/content/neural_profit
[main a29807e] Ventanas X_train/X_valid/X_test y y_train/y_valid/y_test guardadas como .npy
 6 files changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 "3_dise\303\261o_entrenamiento_evaluacion/X_test.npy"
 create mode 100644 "3_dise\303\261o_entrenamiento_evaluacion/X_train.npy"
 create mode 100644 "3_dise\303\261o_entrenamiento_evaluacion/X_valid.npy"
 create mode 100644 "3_dise\303\261o_entrenamiento_evaluacion/y_test.npy"
 create mode 100644 "3_dise\303\261o_entrenamiento_evaluacion/y_train.npy"
 create mode 100644 "3_dise\303\261o_entrenamiento_evaluacion/y_valid.npy"
To https://github.com/GUNAPILLCO/neural_profit.git
 ! [rejected]        main -> main (fetch first)
error: failed to push some refs to 'https://github.com/GUNAPILLCO/neural_profit.git'
hint: Updates were rejected because the remote contains work that you do
hint: not have locally. This is usually caused by another repository pushing
hint: to the same ref. You may want to first integrat